lib/kernel/src/application.erl

```diff
@@ -139,22 +139,40 @@ ensure_all_started(Application, Type) ->
     end.
 
 ensure_all_started(Application, Type, Started) ->
-    case start(Application, Type) of
-	ok ->
-	    {ok, [Application | Started]};
-	{error, {already_started, Application}} ->
-	    {ok, Started};
-	{error, {not_started, Dependency}} ->
-	    case ensure_all_started(Dependency, Type, Started) of
+    case ensure_loaded(Application) of
+	{ok, Name} ->
+	    {ok, Apps} = get_key(Name, applications),
+	    {ok, OptionalApps} = get_key(Name, optional_applications),
+
+	    case ensure_children(Apps, OptionalApps, Type, Started) of
 		{ok, NewStarted} ->
-		    ensure_all_started(Application, Type, NewStarted);
+		    case application_controller:start_application(Name, Type) of
+			ok ->
+			    {ok, [Application | NewStarted]};
+			{error, {already_started, Application}} ->
+			    {ok, NewStarted};
+			{error, Reason} ->
+			    {error, {Application, Reason}, NewStarted}
+		    end;
 		Error ->
 		    Error
 	    end;
 	{error, Reason} ->
 	    {error, {Application, Reason}, Started}
     end.
 
+ensure_children([], _OptionalApps, _Type, Started) ->
+    {ok, Started};
+ensure_children([App | Apps], OptionalApps, Type, Started) ->
+    case ensure_all_started(App, Type, Started) of
+	{ok, NewStarted} ->
+	    ensure_children(Apps, OptionalApps, Type, NewStarted);
+	{error, _AppReason, NewStarted} = Error ->
+	    case lists:member(App, OptionalApps) of
+		true -> ensure_children(Apps, OptionalApps, Type, NewStarted);
```

Comment:
As I said in my docs comment I think this should only be ignored if `App` is not in the code path.

This might be tricky because the same error for not being in the code path is used for not having a `.app` file (so in most cases not compiled because that is when that file is generated): `{error,{"no such file or directory", "App.app"}}`

But even with that caveat I think it is better and more expected to the user to have this whole thing fail if an optional dep fails to start for any reason other than `no such file or directory`.

An example might be a database driver. If I include a certain postgres library as an optional dep and it fails to boot I'd rather know during application start that this has failed than to later get errors when calls to the database are attempted.